### ISSUE TITLE
Support gRPC streaming

### ISSUE THREAD

**Issue by joshcarp** [k6io/k6#2020]
## Feature Description
Support gRPC streaming as described [here](https://k6.io/blog/performance-testing-grpc-services/)

## Suggested Solution (optional)
 The grpc Client currently uses the grpc.Invoke command:
https://github.com/k6io/k6/blob/61f464b99a2d052347fee157eb68a024f5cf7430/js/modules/k6/grpc/client.go#L398

Instead of this, possibly using the return value of `NewStream` [godoc](https://godoc.org/google.golang.org/grpc#ClientConn.NewStream):

https://github.com/grpc/grpc-go/blob/87eb5b7502493f758e76c4d09430c0049a81a557/stream.go#L144

This would look similar to [the generated stream endpoint grpc code](https://github.com/k6io/k6/blob/5e42319785addea437c38b31f92870a10447c8dd/samples/grpc_server/route_guide.pb.go#L594)

```go

func (c *routeGuideClient) ListFeatures(ctx context.Context, in *Rectangle, opts ...grpc.CallOption) (RouteGuide_ListFeaturesClient, error) {
	stream, err := c.cc.NewStream(ctx, &_RouteGuide_serviceDesc.Streams[0], "/main.RouteGuide/ListFeatures", opts...)
	if err != nil {
		return nil, err
	}
	x := &routeGuideListFeaturesClient{stream}
	if err := x.ClientStream.SendMsg(in); err != nil {
		return nil, err
	}
	if err := x.ClientStream.CloseSend(); err != nil {
		return nil, err
	}
	return x, nil
}

```

**Comment by na--** [comment]
Thanks for making this issue!

We'd like to have gRPC streaming, but at least the server streaming and bi-directional streaming parts are somewhat blocked by our lack of event loops in VUs (https://github.com/loadimpact/k6/issues/882). For now, before we have this in the k6 core, someone should probably try to make an [xk6](https://github.com/k6io/xk6) extension first...

**Comment by k30v1n** [comment]
Hi @na-- , I would like to raise the same question again. Apparently, if I get it correctly, the event loops in VUs are now supported and released based on PR https://github.com/grafana/k6/pull/2228 . 

Is there and glimpse of development of this feature?

**Comment by na--** [comment]
Yes! I can't give any promises for when exactly, depends on how we prioritize against a whole bunch of other things, but it's definitely on our short-term roadmap now that we've cleared out the obstacles.

**Comment by k30v1n** [comment]
@na--  Thank you for such a quick response! That is great and very optimistic! `v0.39.0` is really near ❤️ 

To understand a little bit better, would be possible/how to help on the development of this feature? I mean, is there any definition already to how the streaming tests setup will be done (like the ones for [unary calls here](https://k6.io/blog/performance-testing-grpc-services/)) or this is still something open for definition?

**Comment by na--** [comment]
Still TBD. We are still figuring out things when it comes to event loops (e.g. see [this](https://github.com/grafana/k6/tree/master/cmd/integration_tests/testmodules/events) and [this](https://github.com/grafana/xk6-websockets) for some recent experiments) and we haven't started planning how the JS APIs for gRPC streaming should look like exactly. We'll probably research what other JS gRPC libraries use before we design them... :thinking: 

That said, if you want to work on this, go right ahead! As I said, I somewhat doubt we'll manage to get this time in time for v0.39.0, but a good external pull request might speed things up significantly! Just make sure to share your API proposal here for comments, before you've invested a significant amount of time implementing it, otherwise there might be a nasty surprise if we decide not to merge it because of some fundamental misalignment.

**Comment by k30v1n** [comment]
Sure! Thanks for the heads up!

**Comment by nsmith5** [comment]
👋🏻 Hey @na-- and @olegbespalov, are ya'll still looking for some help on this feature (per the `help-wanted` label)? I'm happy to help out how ever necessary as this feature would be super helpful to me!

**Comment by olegbespalov** [comment]
Hey @nsmith5 ! 

I planned to start work on that one, but for sure, community first :wink: So if you want to pick up the issue and work on it, please go ahead :raised_hands: :relaxed:

**Comment by na--** [comment]
@nsmith5, it's great that you want to work on this! As @olegbespalov said, we probably would have picked it up soon, but we'd love to merge your PR instead! :tada: 

Here are some details:
1. Please take a look at the instructions in [CONTRIBUTING.md](https://github.com/grafana/k6/blob/master/CONTRIBUTING.md)
2. Before you start writing any code, please share the JS API you propose to implement. Having a tentative :+1: for the API proposal will make reviewing the code later much easier and increase our chances of actually merging it.
3. If you can, please let us know roughly when you expect to be able to work on this issue. If it's not going to be in the near future, we might prefer to implement it ourselves sooner.
4. We have a new strategy of how we develop and stabilize new JS APIs, see [this part of the v0.39.0 release notes](https://github.com/grafana/k6/blob/master/release%20notes/v0.39.0.md#k6experimental-js-modules) for details. We first try to first implement them as unstable xk6-extensions and iterate on the API, then integrate them in k6 as semi-stable `k6/experimental/whatever` built-in modules and finally release fully stable APIs as `k6/whatever`. In this case, since we already have parts of the gRPC API in k6, we should probably start with the new streaming parts in `k6/experimental`.
5. We still don't have good documentation on how to build asynchronous JS APIs with the new event loops, sorry. But maybe use https://github.com/grafana/xk6-timers and https://github.com/grafana/xk6-websockets as examples.

**Comment by nsmith5** [comment]
Yeah, I don't want to block on ya'll going ahead and implementing yourself. I might try to hack together something to try this soon, but I'm not sure on timescale and have some other priorities a launch that might interrupt. Thanks a bunch on all the pointers on where to learn more and get started. The async js bits in particular I wasn't sure about so I'll take a look at xk6-websockets for sure

**Comment by Dogacel** [comment]
I think the proposed API can look similar to WebSockets & node grpc streaming API (https://grpc.io/docs/languages/node/basics/#streaming-rpcs)

```
var call = client.echoStream(echoRequest);
call.on('data', function(data) {});
call.on('end', function() {});
call.on('error', function(e) {});
call.on('status', function(status) {});
```

**Comment by olegbespalov** [comment]
Hi @nsmith5 !

It seems we are in a position to prioritize this task and start work on that, but since you were part of the discussion above, I'd like to sync with you. Are you okay if we implement this ourselves?  :relaxed:

**Comment by olegbespalov** [comment]
So a minor update here and a proposal for the API. I've checked the nodejs eco-system and to be honest agree with the @Dogacel that our GRPC's stream could look similar to the NodeJS. 

There could be an example of the bidirectional stream.
```js
import grpc from 'k6/experimental/grpc';

const client = new grpc.Client();
client.load([], './proto/employee.proto');

export default () => {
   client.connect('localhost:4500', {
      plaintext: true
   });
   
   // a client opens a bidirectional stream
   const metadata = {};
   const stream = client.newStream('employee.Employee/generateReport', metadata);

  stream.on('data', (message) => {
     // server send data, processing...
  });

   stream.on('end', () => {
      // server has finished sending
   });

   stream.on('error', (err) => {
      // process the err
   });

   // client writes something to the stream
   stream.write({message: 'foo'});
   stream.write({message: 'bar'});
   
   // client ends the stream
   stream.end();
  
  client.close();
};
```

What do you think?

**Comment by allansson** [comment]
A few thoughts on the proposed API.

> ```js
> import grpc from 'k6/experimental/grpc';
> 
> const client = new grpc.Client();
> ```

 My suggestions would be to do:

```js
import GrpcClient from 'k6/experimental/grpc' // Default export is the class constructor
import { Client } from 'k6/experimental/grpc' // Named exports for everything that the module exports
import * as grpc from 'k6/experimental/grpc' // Allow namespacing via wildcard imports
```



The reason NodeJS has the `grpc` namespace is that Node (mostly) uses `require`(i.e. CommonJS). With ES modules there's a lot more freedom.

---

> ```js
> client.load([], './proto/foo.proto');
> ```

If passing the client some definitions is a requirement, might as well offer an option to pass them in the constructor 
(I'm not sure what the empty array does though, so the example is probably not accurate)

```js
new Client({
  definitions: [
    './proto/foo.proto'
  ]
})
```

---

> ```js
> client.connect('localhost:4500', {
>    plaintext: true
>  });
>
> client.close();
> ```

`open` + `close` or `connect` + `disconnect`?

---

> ```js
> const stream = client.newStream('foo.BarService/someMethod', metadata);
> stream.end();
> ```

`newStream` feels weird to me when JS has the `new` keyword: `new Stream(client)`.

An alternative is `startStream`/`beginStream` to mirror the `end` method.

**Comment by olegbespalov** [comment]
For the convenience extracted the proposed API as the separated messaged

```js
import { Stream } from 'k6/experimental/grpc'

// ...
// existing initialization and connection of the client from `k6/net/grpc`
// ...

// any message that you want to send to the server with the opening a stream
const message = {text: 'hello world'};

// params represent the additional params that we could configure
// it could be the metadata, tags and extended in the future
const params = {
    tags: {
       my_tag: 'my_value'
    }
};

// this constructs the stream
// Stream(client, method, [,request], [, params])
// - client - already initialized and connected client with the loaded definitions
//   ideally it should be the same client that currently exists in k6/net/grpc
//   later we could also accept the client that constructed using new GRPC api
// - method - a gRPC method url to invoke.
// - request - a canonical request object, as-per the Protobuf JSON Mapping.
// - params - an object with additional params that could be passed to the stream.
const stream = new Stream(client, 'employee.Employee/generateReport', message, params)


// an `on` method could be used to register the event handlers
// for the server streams

// a `data` event is triggered when the server sends the data
// you could register multiple handlers for the same event
stream.on('data', (message) => {
  // server send data, processing...
})

// an `end` event is triggered when the server has finished sending the data
// you could register multiple handlers for the same event
stream.on('end', () => {
  // server has finished sending
})

// an `error` event is triggered when an error occurs
// you could register multiple handlers for the same event
stream.on('error', (err) => {
  // process the err
})


// a `write` and `end` methods could be used for the client streams

// write writes data to the stream
stream.write({ message: 'foo' })
stream.write({ message: 'bar' })

// signal the server that the client has finished sending the data
stream.end()
```

**Comment by andreea-anghel** [comment]
Do you have a rough estimate of when this feature will be available? Thanks!

**Comment by olegbespalov** [comment]
Hi @andreea-anghel 

This feature is prioritized to land in the following `v0.44.0` release, which should happen in mid of April. For sure, it's a soft commitment, but we're doing our best.

**Comment by Dogacel** [comment]
@olegbespalov do you need any contribution regarding development?

**Comment by olegbespalov** [comment]
Hey @Dogacel ! Thanks for offering help :relaxed:  For now all good :+1:

**Comment by m-guesnon-pvotal** [comment]
Hi!

I see the target milestone has been pushed.  Is the milestone a target for GA release or experimental?

**Comment by olegbespalov** [comment]
Hi @m-guesnon-pvotal and the other topic-followers!

Yeah, it's worth updating here. 

Unfortunately, the GRPC streaming won't be part of the `v0.44.0` release. There were a few things that affected pace :cry: 

The whole GPRC logic is extracted as https://github.com/grafana/xk6-grpc. We're working to make GRPC streaming happen. So in the following weeks, it will be available as the k6's extension, so it will be possible to try it by building the custom binary. And it's going to be included in the k6 as an experimental module in `v0.45.0`.

I'm also going to make the status updates here once the things from above will happen.

Thanks for your patience and sorry for the delay.

**Comment by olegbespalov** [comment]
Here is a minor status update.

The PR with an experimental GRPC module is there #3107, and we're planning to merge it soon, making a new experimental GRPC module available in the [k6 v0.45](https://github.com/grafana/k6/milestone/33).

The new experimental module extends the current k6's GRPC module, that way that the end user could just replace the module import:

```diff
-import { Client } from 'k6/grpc';
+import { Client } from 'k6/experimental/grpc';
```

However, I must stress that this could also change in a feature since we have some discussions about improving the Client's API https://github.com/grafana/k6/issues/2020#issuecomment-1458450772

We continue improving the extension, and hopefully, it could replace the existing k6 grpc module at some point.

The documentation of the API will be available soon (but definitely before the release), and for now, you can check the straightforward examples:
* https://github.com/grafana/k6/blob/0167ce449c38b745bf8e6448582b84fdb0972c63/examples/experimental/grpc/client_streaming.js
* https://github.com/grafana/k6/blob/0167ce449c38b745bf8e6448582b84fdb0972c63/examples/experimental/grpc/server_streaming.js

If you face any issues or have some improvements suggestions, please open the issue in the extension's repository: https://github.com/grafana/xk6-grpc/issues
